### PR TITLE
[shardformer] shardformer support jit fused operator.

### PR DESCRIPTION
**Pull request by flybird1111**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

#4235 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

bloom support jit fused operator.

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Comment by flybird1111** [comment]
image
All tests have passed.

File: colossalai/shardformer/modeling/bert.py [hunk 1]
```diff
@@ -109,5 +111,24 @@ def bert_flash_attention_forward(
...
             outputs = outputs + (past_key_value,)
         return outputs
-    
-    return bert_flash_attention_forward
\ No newline at end of file
+
+    return forward
+
+
+def _get_jit_fused_output_forward():
```
Comment (ver217) [inline colossalai/shardformer/modeling/bert.py:118]: This closure does not have any local variables. What's the reason to use closure?

File: colossalai/shardformer/modeling/bert.py [hunk 2]
```diff
@@ -109,5 +111,24 @@ def bert_flash_attention_forward(
...
+        return hidden_states
+
+    return forward
+
+
+def get_jit_fused_bert_self_output_forward():
+    return _get_jit_fused_output_forward()
+
+
+def get_jit_fused_bert_output_forward():
```
Comment (ver217) [inline colossalai/shardformer/modeling/bert.py:133]: These two functions call the same function but their name are different. It's confusing.

File: colossalai/shardformer/modeling/bloom.py [hunk 3]
```diff
@@ -71,15 +71,15 @@ def build_bloom_alibi_tensor(self, attention_mask: torch.Tensor, num_heads: int,
...
 
     try:
         from xformers.ops import memory_efficient_attention as me_attention
     except:
         raise ImportError("Error: xformers module is not installed. Please install it to use flash attention.")
-    from transformers.models.bloom.modeling_bloom import dropout_add
 
-    def bloom_flash_attention_forward(
+    def forward(
         self,
```
Comment (ver217) [inline colossalai/shardformer/modeling/bloom.py:83]: self should have type hint

File: colossalai/shardformer/modeling/bloom.py [hunk 4]
```diff
@@ -114,30 +115,171 @@ def bloom_flash_attention_forward(
...
 
         return outputs
-        
-    return bloom_flash_attention_forward
\ No newline at end of file
+
+    return forward
+
+
+def get_jit_fused_bloom_attention_forward():
```
Comment (ver217) [inline colossalai/shardformer/modeling/bloom.py:157]: Same problem

File: colossalai/shardformer/modeling/bloom.py [hunk 5]
```diff
@@ -114,30 +115,171 @@ def bloom_flash_attention_forward(
...
+
+    def forward(
+        self,
```
Comment (ver217) [inline colossalai/shardformer/modeling/bloom.py:160]: Same problem

File: colossalai/shardformer/modeling/bloom.py [hunk 6]
```diff
@@ -114,30 +115,171 @@ def bloom_flash_attention_forward(
...
+        outputs = (output_tensor, present)
+        if output_attentions:
+            outputs += (attention_probs,)
+
+        return outputs
+
+    return forward
+
+
+def get_jit_fused_bloom_mlp_forward():
```
Comment (ver217) [inline colossalai/shardformer/modeling/bloom.py:253]: Same problem

File: colossalai/shardformer/modeling/bloom.py [hunk 7]
```diff
@@ -114,30 +115,171 @@ def bloom_flash_attention_forward(
...
+
+    def forward(self, hidden_states: torch.Tensor, residual: torch.Tensor) -> torch.Tensor:
```
Comment (ver217) [inline colossalai/shardformer/modeling/bloom.py:255]: Same problem

File: colossalai/shardformer/modeling/jit.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+    def self_dropout_add(self, x: torch.Tensor, residual: torch.Tensor, prob: float, training: bool) -> torch.Tensor:
+        return dropout_add(x, residual, prob, training)
+
+    return self_dropout_add
+
+
+def get_jit_fused_dropout_add_func():
+
+    from colossalai.kernel.jit import bias_dropout_add_fused_inference, bias_dropout_add_fused_train
```
Comment (ver217) [inline colossalai/shardformer/modeling/jit.py:16]: Does JIT ops have any compatibility problem?

Comment (flybird1111) [reply colossalai/shardformer/modeling/jit.py:16]: > Does JIT ops have any compatibility problem?

I haven't encountered any compatibility problem with the JIT operators so far.

Comment (flybird1111) [reply colossalai/shardformer/modeling/jit.py:16]: > Does JIT ops have any compatibility problem?

Updated

File: colossalai/shardformer/modeling/bloom.py (hunk 7 above)
Comment (flybird1111) [reply colossalai/shardformer/modeling/bloom.py]: > Same problem

Updated.

File: colossalai/shardformer/modeling/bloom.py (hunk 3 above)
Comment (flybird1111) [reply colossalai/shardformer/modeling/bloom.py]: > self should have type hint

updated.

File: colossalai/shardformer/modeling/bert.py (hunk 1 above)
Comment (flybird1111) [reply colossalai/shardformer/modeling/bert.py]: > This closure does not have any local variables. What's the reason to use closure?

updated.

File: colossalai/shardformer/modeling/bert.py (hunk 2 above)
Comment (flybird1111) [reply colossalai/shardformer/modeling/bert.py]: > Member

Updated.

**Review by ver217 — APPROVED** [review]

**Pull request merged**